PAS-7: Write CDK templates for infrastructure

BraidenCutforth opened the pull request.

BraidenCutforth committed:
Move infrastructure to new location

BraidenCutforth committed:
heavy wip

BraidenCutforth committed:
CDK-TEST: WIP

- Testing out the CDK
- Complete the cloudfront setup
- Separated out the Mail record sets
- Parameterized all variables needed if you were to deploy a different
website

BraidenCutforth committed:
PAS-7 and More

- Re-created entire infrastructure using aws-cdk
- Removed the user pool from the frontend code
- Created a custom resource for storing parameters from stacks made in
different regions

BraidenCutforth committed:
PAS-7: Remove passr direct references in CDK

- Add cache to buildspec
- Cleanup CDK templates

BraidenCutforth committed:
PAS-109: Remove old infrastructure

daelynj commented:
~~Is there any reason we're keeping old infrastructure? We have powerful tools like version control that let us keep an infinitely long history of code, even after it's been removed.~~

Missed the last commit.

daelynj commented on infrastructure/bin/infrastructure.ts:
```diff
@@ -0,0 +1,44 @@
+import { FrontendStack } from '../lib/frontend';
+import { MailStack } from '../lib/mail';
+import { ProjectPrefix } from '../lib/variables';
+const app = new cdk.App();
```
Probably good to separate this instantiation from the imports list.

BraidenCutforth replied:
Done.

daelynj commented on infrastructure/lib/variables.ts:
```diff
@@ -0,0 +1,8 @@
+export const DomainName = 'passr.ca';
+export const MXSPFRecords = ['"v=spf1 include:spf.improvmx.com ~all"'];
+export const MXReconds = ['10 mx1.improvmx.com', '20 mx2.improvmx.com'];
+export const ProjectPrefix = 'passr';
```
These feel like constants to me, and I feel like they should be declared and denoted as such. My suggestion whenever I see a file with constants in it like this in JS is to structure them like this:
```js
// infrastructure/lib/constants.ts

const DOMAIN_NAME = 'passr.ca';

const FRONTEND_MAIN_BRANCH = 'master';

const FRONTEND_REPO = 'Passr';

const GITHUB_SECRET_NAME = 'github';

const MX_RECORDS = ['10 mx1.improvmx.com', '20 mx2.improvmx.com'];

const MXSPF_RECORDS = ['"v=spf1 include:spf.improvmx.com ~all"'];

const PROJECT_PREFIX = 'passr';

const REPO_OWNER = 'VikeLabs';

export default {
  DOMAIN_NAME,
  FRONTEND_MAIN_BRANCH,
  FRONTEND_REPO,
  GITHUB_SECRET_NAME,
  MX_RECORDS,
  MXSPF_RECORDS,
  PROJECT_PREFIX,
  REPO_OWNER,
};
```
```js
// usage

import constants from 'infrastructure/lib/constants';

console.log(constants.DOMAIN_NAME);
```
Note the constant names are alphabetized.

There are a few reasons for doing all of this:
A dev 250 lines deep into a file who sees `RepoOwner` (because it's explicitly imported) needs to spend a minute trying to figure out where `RepoOwner` came from, but if they see `constants.REPO_OWNER` they immediately know the variable is a constant and comes from the `constants` import.

Another reason is that you limit the export type of the entire module to explicitly being an object, rather than hail marry exports where any type could be imported/exported from the file. Any time you can avoid multiple return types from a function or module is a good time.

Another reason to do this is because it encourages you to separate constants into modules of separate concerns. For instance in that constants file we have some domain constants and github constants. So while not necessary right now, breaking constants into files separated by concerns lets you import things like
```js
import githubConstants from 'infrastructure/lib/constants/github';
import domainConstants from 'infrastructure/lib/constants/domain';
```

These are some things to think about for larger projects or if you think these constants could get larger.

BraidenCutforth replied:
Yeah good call. 

I don't think I understand this:

> Another reason is that you limit the export type of the entire module to explicitly being an object, rather than hail marry exports where any type could be imported/exported from the file. Any time you can avoid multiple return types from a function or module is a good time.

I always really like named exports, and many libs use them but yeah idk. The thing I don't understand is having a module return multiple types... My thought is what does it matter if you export 3 different types vs returning an object that contains 3 different types? Wrapping a 'hail marry export' in an object doesn't really make it any cleaner?

daelynj replied:
It's splitting hairs and comes down to opinion mostly. By always exporting an object we're introducing encapsulation to the module, a public and private interface if you will. We can argue that named exports also provide encapsulation, which they do, but there can be a bit of misdirection when using them in a constants file like this.

To my example of being 250 lines deep into a file and seeing the variable `RepoOwner`, to know what this is and where it comes from, a developer needs to CTRL+F for `RepoOwner`, or use some editor 'find occurences of' functionality or something, then investigate where it comes from. A small change that would make this immediately clearer would be to rename it `REPO_OWNER` which immediately screams "constant", as constants should always be capitalized and snake_case to begin with. But again, where is it coming from. By using default exports in the case of a constants file when a developer sees `domainConstants.REPO_OWNER` 250 lines into a file it's immediately apparent where it's coming from and what it is. This saves the 30 seconds of misdirection.

This is splitting hairs because what is 30 seconds to one or two people, but I suppose on a team of 10 to 100 people, 30 seconds each is a lot of money.

-------------

To the point of exporting 3 different types vs. returning an object that contains 3 different types, I would maybe say don't think about what is in the Object, just consider the fact that when importing any constants file, you _always_ get an Object. This is an honest and consistence interface. This is also splitting hairs because with one or two people it really doesn't matter but when you have a `constants/` directory with 100+ files, the consistency is important.

Months ago at my work we were performing major refactors and we had a random assortment of export types in our constants files, some were named, some weren't, some had both, and the inconsistency was so incredibly rage inducing any time we went to work with constants in different files etc. Having a single API for constants definitely made our code look so much better and made importing constants consistent and clean.

-----------

Again these things aren't really that important right now but I figure I'll point them out because I like to sound like I know what I'm doing 😄

daelynj commented on infrastructure/lib/frontend.ts:
```diff
@@ -0,0 +1,158 @@
+				zoneName: hostedZoneName,
+				hostedZoneId,
+			}
+		);
```
I'm noticing a trend between all of these classes in this PR in that we're operating at a single level of abstraction in all of them. This is clear when you try to discern the responsibilities that the `FrontendStack` class has for example, setting up these:
1. `hostedZone`
2. `certificiate`
3. `cloudFront`

as well as some other things happening in this class.

This is problematic if we think these classes are ever going to get bigger or be extended with additional concerns. If we know these classes will be left as is or only modified, then maybe this is okay.

BraidenCutforth replied:
I am not exactly sure what you mean by this, but I am following the CDK documentation on building a stack. Maybe they have better docs for what to do as your infrastructure grows, but I am trying to limit the amount of things in a stack, by grouping things that are coupled together. I think this might just be a result of how infrastructure is, but interested to see how I could improve on this.

daelynj replied:
Yeah so when I mention operating at a single level of abstraction I mean there are a lot of implementation details in this file, giving it tons of responsibilities; we're at a very shallow level. If we lift ourselves up to a higher level of abstraction and bury these implementation details in factories / our own classes wrapping the CDK functions etc. we hide all of these implementation details that this file shouldn't necessarily be concerned about, which brings us closer to our class having a single responsibility.

What we have here is great for a one-and-done 'define it once' scenario, but to my point of "this is problematic if we think these classes are ever going to get bigger", having a shallow level of abstraction like this leads to big confusing files and tons of implementation details can be problematic.

daelynj commented on infrastructure/lib/mail.ts:
```diff
@@ -0,0 +1,55 @@
+				parameterName: HostedZoneIdParam,
+				region: 'us-east-1',
+			}
+		);
```
There's actually a fair bit of code duplication between these classes.

Let's say for example we need to change the region all of our stack is using, `us-east-1`. How many places would we need to make this change? I believe I count 6. What happens if one is missed?

It wouldn't be terribly difficult to take this one level of abstraction deeper and create a factory wrapper of sorts for example with all the `hostedZone` code. We can do the same with lots more of this code too.

I don't even know how to write gross class-oriented JS so I'm just going to do it with functional syntax lol

```js
// infrastructure/lib/factories/hostedZone

const hostedZoneFactory = (stackInstance, ...) => { // stackInstance is the 'this' keyword passed in
  const idReader = ...
  const nameReader = ....

  return {
    idReader,
    nameReader,
    ...
  }
}
```

```js
// usage

import hostedZoneFactory from 'infrastructure/lib/factories/hostedZone';

export class MailStack extends cdk.Stack {
  constructor(...) {
    ...
    const hostedZone = hostedZoneFactory(this, ...);

    // use hostedZone.idReader wherever needed, etc.
    
   // alternatively
    const { idReader, nameReader, ... } = hostedZoneFactory(this, ...);
```

Now we have a hostedzone "class" that we can use throughout these classes at a higher level of abstraction and we only need to make updates to it in one place.

On another note, things like `'us-east-1'` should be placed in a constants file so again there is only ONE place for them to be modified.

```js
region: regionConstants.REGION,
```

BraidenCutforth replied:
Yeah this is a great design, and I am going to implement it. I think I got a bit lazy after reading so much documentation just to get it to work 😂

daelynj replied:
To the point of having different levels of abstraction, adding factories like this introduces a new level, where the `frontendStack` class would be interacting with these factories now, which would in turn be interacting with the CDK stack. With this example the idea becomes clear I hope, we have
```
// before
FrontendStack class
 - HostedZone from CDK
 - CloudFront from CDK
 - AnythingElse from CDK

// after
FrontendStack class
  - HostedZoneFactory
      - HostedZone from CDK
  - CloudFrontFactory
      - CloudFront from CDK
  - AnythingElseFactory
      -AnythingElse from CDK
```

We hide the implementation details in the factories, or classes, whatever you want them to be. This removes responsibilities from the FrontendStack class. You can say the factories are operating at a lower level of abstraction and the FrontendStack class is operating at a higher level of abstraction. The more levels we introduce the more we separate concerns and create reusable code, though this can be obviously taken too far.

Hope this is clear!

BraidenCutforth committed:
PAS-7: Standardize Parameters

- Move all constants to a constants file and export
- Create a generic factory for importing variables from other stacks
- Implement the generic factory for existing global variables

isaiahdoyle commented on infrastructure/lib/frontend.ts:
```diff
@@ -0,0 +1,133 @@
+			certificateArn
+		);
+
+		const cloudfrontDev = new cloudfront.CloudFrontWebDistribution(
+			this,
+			'FrontendDevDistribution',
+			{
+				originConfigs: [
+					{
+						s3OriginSource: {
+							s3BucketSource: props.devSiteBucket,
+							originAccessIdentity: props.devOAI,
+						},
+						behaviors: [{ isDefaultBehavior: true }],
+					},
+				],
+				defaultRootObject: 'index.html',
+				errorConfigurations: [
+					{
+						errorCode: 404,
+						responseCode: 200,
+						responsePagePath: '/index.html',
+					},
+				],
+				priceClass: cloudfront.PriceClass.PRICE_CLASS_100,
+				viewerCertificate: cloudfront.ViewerCertificate.fromAcmCertificate(
+					certificate,
+					{
+						aliases: [`dev.${constants.DOMAIN_NAME}`],
+					}
```
Geez, Prettier really went ham on this one

isaiahdoyle commented on infrastructure/lib/param-factories/certificate-arn.ts:
```diff
@@ -0,0 +1,9 @@
+
+const CertificateArn = GenericParamFactory(
+	constants.CERTIFICATE_ARN_PARAM,
+	'us-east-1'
```
Is this meant to be `east`?

BraidenCutforth replied:
So unfortunately we are forced to use this region when making a certificate for cloudfront. Not a big deal in terms of speed or anything, just means we split some stuff across regions since I think it is still best to host other stuff in oregon (since it is closer to us)

isaiahdoyle commented on infrastructure/lib/param-factories/hosted-zone-id.ts:
```diff
@@ -0,0 +1,9 @@
+
+const HostedZoneId = GenericParamFactory(
+	constants.HOSTED_ZONE_ID_PARAM,
+	'us-east-1'
```
Same here - is there a west alternative? Or is east better in this case?

isaiahdoyle reviewed (approved):
Code looks good!!! I'm very new to this kind of work so I'm unsure if there were any subtle logic/syntax errors although I had a few things to point out. :]

seiyaterada reviewed (approved):
Looks good!
